libs/mrwebrtc-unityplugin/include/export.h

```diff
@@ -0,0 +1,16 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License. See LICENSE in the project root for license
+// information.
+
+#pragma once
+
+// P/Invoke uses stdcall by default. This can be changed, but Unity's IL2CPP
+// does not understand the CallingConvention attribute and instead
+// unconditionally forces stdcall. So use stdcall in the API to be compatible.
+#if defined(MR_SHARING_WIN)
+	#define MRS_API __declspec(dllexport)
+	#define MRS_CALL __stdcall
+#elif defined(MR_SHARING_ANDROID)
+	#define MRS_API __attribute__((visibility("default")))
+	#define MRS_CALL __attribute__((stdcall))
+#endif
```

Comment:
This looks wrong, `MR_SHARING_WIN` (and `_ANDROID`) are used by mrwebrtc to turn on the function export, so if you use the same macro in mrwebrtc-unityplugin the compiler will compile in and export the same functions twice.

At the moment `export.h` in mrwebrtc lacks a configuration where the API is compiled as `dllimport`, so if you want to correctly use the mrwebrtc functions in mrwebrtc-unityplugin you should rewrite that file (and this file, with different macros) so that it handles that case, like the example in https://stackoverflow.com/a/2164853/6705316.

A quicker-and-dirtier alternative is not to include _any_ header from mrwebrtc in mrwebrtc-unityplugin and redefine the types you need (`mrsRemoteVideoTrackHandle`, `mrsResult`). Even in that case you should rename the macros in this file (`MR_SHARING_WIN`, `MRS_API`, `MRS_CALL`) to something different to avoid confusion.

Reply:
This file I meant to delete. I switched it to using the export.h imported from the actual mrwebrtc project. Does doing it that way still cause any issue?

Reply:
Using the same file or a copy-paste doesn't make any difference if you are still using the same macros for both DLLs. unityplugin is compiled as a different DLL from mrwebrtc, so it should use different macros. If they all use `MRS_API`, when you compile unityplugin:
- the functions that are defined in unityplugin are compiled as MRS_API = dllexport (correct)
- any of the API functions in mrwebrtc included by unityplugin will also be compiled as MRS_API = dllexport (wrong! unityplugin doesn't contain its definition, only the declaration, so the linker will fail; should be dllimport instead, because it is importing the function from mrwebrtc)

I am not sure how it is working now, I suppose that since unityplugin has an exports.def, all `__declspec(export)` are ignored (and `__declspec(dllimport)` is optional, so the mrwebrtc functions get imported correctly anyway). But this is bug-prone and confusing, so it should be done the correct way. Since you don't really need `dllimport` after all, I guess a simple fix is this:
- re-add export.h
- rename all the macros to something unique to unityplugin (e.g. `MR_SHARING_`, `MRS_` -> `MRWEBRTC_UNITYPLUGIN_`)
- use the renamed macros in `mrwebrtc-unityplugin/include/api.h` and in the preprocessor definitions in the `vcxproj`

Reply:
Thank you for explaining all of that. Setting up dll exports like this on C++ libraries to go into unity is one of the areas I am not the most knowledgeable about.

I deleted the exports.def file so that it goes off al of __declspec(export) calls. But it still looks like it is correct in what it exports. Running dumpbin I get:
```
1    0 000010F0 OnGraphicsDeviceEvent = OnGraphicsDeviceEvent
2    1 00001260 UnityPluginLoad = UnityPluginLoad
3    2 000012B0 UnityPluginUnload = UnityPluginUnload
4    3 00001330 mrsNativeRenderer_Create = mrsNativeRenderer_Create
5    4 00001450 mrsNativeRenderer_Destroy = mrsNativeRenderer_Destroy
6    5 000015B0 mrsNativeRenderer_DisableRemoteVideo = mrsNativeRenderer_DisableRemoteVideo
7    6 00001470 mrsNativeRenderer_EnableRemoteVideo = mrsNativeRenderer_EnableRemoteVideo
8    7 000015D0 mrsNativeRenderer_GetVideoUpdateMethod = mrsNativeRenderer_GetVideoUpdateMethod
9    8 000012E0 mrsNativeRenderer_SetLoggingFunctions = mrsNativeRenderer_SetLoggingFunctions   
```

Does this mean that the unityplugin.dll is not compiling in any methods from mrwebrtc.dll? I suspect this is so because as it currently is, you must put both the unityplugin.dll and mrwebrtc.dll into Unity. The unityplugin.dll dynamically links to the mrwebrtc.dll

But, I will rename unity macros as you suggest, I assume that is more correct even if it is dynamically linking.

Reply:
However I have found that I still need `MR_SHARING_WIN` defined in the preprocessor definitions for the mrwebrtc-unityplugin project. Is that an issue?

Reply:
I am a bit surprised that it works - it might be that the linker figures out that the functions are imported from mrwebrtc on its own.

You should rename `MR_SHARING_WIN` and `MR_SHARING_ANDROID` too, they are supposed to be specific to the DLL too.

> Setting up dll exports like this on C++ libraries to go into unity is one of the areas I am not the most knowledgeable about.

I have always found this to be surprisingly badly documented, the MS docs explain what things do but fail IMHO to give a comprehensive usage picture. You can find a few examples on SO, e.g. https://stackoverflow.com/questions/538134/exporting-functions-from-a-dll-with-dllexport, maybe that will give you a few more info.